### PR TITLE
Update update-adtools-sdk-nonfree

### PR DESCRIPTION
**Pull request by AV00**
Changes to fetch the latest SDK and install Documentation

### CODE REVIEW

**Comment by sba1** [comment]
Thanks for the PR. I have a few things to remark. First, I think that downloading always the latest version will result in irreproducible builds of the compiler packages. Also, when installing the same Debian package (which uses this script) the result of the installation process should be always the same. This would be not true if always the latest version would be used. Also keep in mind, if for instance the structure of the SDK changes, then the script would fail.

So it should be always a fixed version and we need to bump it whenever a new SDK is released (this doesn't happen that often anyway). Could you change that? What would be a nice improvement would be to have a parameter that allows one to specify the version of the SDK to download and perhaps as a special input that the lastest one should be pulled.

In the future I hope that we can include relevant parts of the SDK directly in the Debian package, but this is not possible for now.

**Comment by AV00** [comment]
On Sat, February 11, 2017 11:11 pm, Sebastian Bauer wrote:
>
Thanks for the PR. I have a few things to remark. First, I think that
> downloading always the latest version will result in irreproducible
builds
> of the compiler packages.

I'm not sure I
understand what you mean: the compiler packages you build for debian
aren't related to the content of the SDK. The sdk is installed after, by
this script.

> Also, when installing the same Debian
package
> (which uses this script) the result of the installation
process should be
> always the same. This would be not true if
always the latest version would
> be used.

I disagree
and I have one example to back it up: the Flash player installer (as a
matter of fact, I took it as an inspiration). It works in a similar way:
it finds out what's the latest version and downloads it. The package is
named flashplugin-nonfree in case you want to check it out. It WILL create
different results depending on which is the latest flash plugin.

Your assessment would be true if we packaged the entire SDK as a debian
package. But we aren't, we're only packaging a downloader script.

> Also keep in mind, if for instance the structure of the SDK
> changes, then the script would fail.

Then we should do
like you ask below...

> So it should be always a fixed
version and we need to bump it whenever a
> new SDK is released
(this doesn't happen that often anyway). Could you
> change that?
What would be a nice improvement would be to have a parameter
>
that allows one to specify the version of the SDK to download and
perhaps
> as a special input that the lastest one should be
pulled.

First of all, I've already changed the
"absolute" filenames to patters, for clib and newlib, so there's
at least some flexibility if the structure changes.

Second...
I'll have to think about your idea of "one update of the script per
sdk release". I guess I should play with debconf to ask which SDK to
install, and so on.


> In the future I hope that we can
include relevant parts of the SDK
> directly in the Debian
package, but this is not possible for now.

Which brings up
another question (copyright and such) regarding my other proposal, using
my tool that converts to doxygen the autodocs. I wonder if it'd be allowed
to distribute the derived work.


Thanks for the ideas,
anyway.

**Comment by sba1** [comment]
> I'm not sure I
> understand what you mean: the compiler packages you build for debian
> aren't related to the content of the SDK. The sdk is installed after, by
> this script.

The build of adtools-gcc depends on adtools-sdk. 

> I disagree
> and I have one example to back it up: the Flash player installer (as a
> matter of fact, I took it as an inspiration). It works in a similar way:
> it finds out what's the latest version and downloads it. The package is
> named flashplugin-nonfree in case you want to check it out. It WILL create
> different results depending on which is the latest flash plugin.

However, AFAIK no build of "official" packages depend on the flash plugin, so the case is a little bit different. Note that we don't need to make it in the way how it us currently done (because it is not perfect anyway). For instance, there could we an adtools-sdk-53.24 Debian package and an adtools-sdk meta package that points to the lastest one. The script is then provided in adtools-common or such.

Generally, it is a good idea to have reproducible builds. Not only the adtools package depend on it but also many CI environments.

> Which brings up
> another question (copyright and such) regarding my other proposal, using
> my tool that converts to doxygen the autodocs. I wonder if it'd be allowed
> to distribute the derived work.

We cannot distribute this kind of stuff at this time, however, we could provide the base tooling for this, so this kind of work is much appreciated. We probably should have a dedicated package for it or an put it in the before mentioned adtools-common.

**Comment by AV00** [comment]
On domenica 12 febbraio 2017 10:02:41 CET Sebastian Bauer wrote:

> The build of adtools-gcc depends on adtools-sdk.

Yes, got it. Though I could argue that the entry points involved on running 
GCC and binutils aren't going to change between one SDK or the other, in 
principle you're right. I hadn't thought about this aspect, sorry.

> However, AFAIK no build of "official" packages depend on the flash plugin,

True...

> so the case is a little bit different. Note that we don't need to make it
> in the way how it us currently done (because it is not perfect anyway). For
> instance, there could we an adtools-sdk-53.24 Debian package and an
> adtools-sdk meta package that points to the lastest one. The script is then
> provided in adtools-common or such.

Ok, I guess it's more work but it's a cleaner solution and very "Debian-like".

> Generally, it is a good idea to have reproducible builds. Not only the
> adtools package depend on it but also many CI environments.

Understood.

> We cannot distribute this kind of stuff at this time, however, we could

(Side note: I personally believe that anything that could restrict the  
freedom of developers on what's left of this platform is stupid. But I'm not 
in charge)

> provide the base tooling for this, so this kind of work is much
> appreciated. We probably should have a dedicated package for it or an put
> it in the before mentioned adtools-common.

I like the idea of making other developers life easier, but due to what 
happened in the past, I don't want Hyperion to use my work. If I were to 
contribute the source here, I'd lose control of it. So I'd rather only 
distribute the processed file, but as you said that can't be done.

Back on topic: what now ?

I'm not too much of a ninja for debian packaging. So if you don't mind I'd 
leave up to you building the meta package. Just tell me what the requirements 
of the script would be. You mentioned that it should support downloading 
different versions, right ? And have a switch instead to just grab the latest 
one , like it's doing now ?
Would that be all ?


-- 
Bye, A.

**Comment by jens-maus** [comment]
@AV00 I agreee with @sba1 that the makefile should be fixed to a certain SDK release and not always downloading the latest one or otherwise this would result in build problems as soon as a new OS4 SDK is released and this repository is not updated to support this SDK.

**Comment by sba1** [comment]
> (Side note: I personally believe that anything that could restrict the
> freedom of developers on what's left of this platform is stupid. But I'm not
> in charge)

I couldn't agree more. But I'm not in charge either. I will however bring up this topic again.

> I like the idea of making other developers life easier, but due to what
> happened in the past, I don't want Hyperion to use my work.

Okay, understood, though this is a pitty because this work needs to be redone by someone else then. It is very useful I think. But yes, everybody could use that work in some form or another if you contribute it here.

You are welcome in case  you change your opinion regarding the this issue.

> If I were to
> contribute the source here, I'd lose control of it. So I'd rather only
> distribute the processed file, but as you said that can't be done.

Okay, but honestly, I wouldn't include only the processed files in context of this project even if there wouldn't be any copyright issues without haven an open source variant of the tool.

> I'm not too much of a ninja for debian packaging. So if you don't mind I'd
> leave up to you building the meta package. Just tell me what the requirements
> of the script would be. You mentioned that it should support downloading
> different versions, right ? And have a switch instead to just grab the latest
> one , like it's doing now ?
> Would that be all ?

Yes, sounds good. I think it is enough to specify which version to download and an option to download the lastest one. The remaining one is some Debian magic which is better done outside of the script (e.g., in the post install script, where also the debconf stuff can happen). I can do this if you wish. Note however I like to license this script under a very liberal license. I haven't made up my mind which one, probably something like MIT or BSD.

There are few other ideas like comparing the SHA1 of the downloaded package but these are separate issues.

**Comment by AV00** [comment]
On lunedì 13 febbraio 2017 11:32:34 CET Sebastian Bauer wrote:

> Okay, understood, though this is a pitty because this work needs to be
> redone by someone else then. It is very useful I think. But yes, everybody
> could use that work in some form or another if you contribute it here.

It's useful, I can browse using firefox/timberwolf the documentation, I have 
full-text search done by the browser and cross reference. The only thing that 
doesn't work is referencing the system includes, I've fiddled with the options 
of Doxygen but couldn't get it to work.
Well, that and the tons of mistakes in the original autodocs (misspelled 
function names and other stuff). I can send you privately a sample. 

> You are welcome in case  you change your opinion regarding the this issue.

There's a lot behind my decision, and unless the "a lot" changes, I'm not 
going to change my mind.

> Okay, but honestly, I wouldn't include only the processed files in context
> of this project even if there wouldn't be any copyright issues without
> haven an open source variant of the tool.

Yes, I agree on the principle.

> Yes, sounds good. I think it is enough to specify which version to download
> and an option to download the lastest one.

Ok, I've spent few hours on the issue (starting to know bash well and hate 
it), I need some pointers: what arguments should I prepare for ?
Currently you had "--install" and "--uninstall". So how to specify version/
revision or "latest" (I suppose only on install) ?
Something like "--latest" and "--version vv.rr"  ? Or separate version and 
revision like "--version vv --revision rr".

> The remaining one is some Debian
> magic which is better done outside of the script (e.g., in the post install
> script, where also the debconf stuff can happen). I can do this if you
> wish.

Yes please.

> Note however I like to license this script under a very liberal
> license. I haven't made up my mind which one, probably something like MIT
> or BSD.

Sure. Anyway mine is derived work on this one, so apply whichever license you 
seem fit.

> There are few other ideas like comparing the SHA1 of the downloaded package
> but these are separate issues.

Yes, let's close this one first.

-- 
Bye, A.

**Comment by sba1** [comment]
> Ok, I've spent few hours on the issue (starting to know bash well and hate
> it), I need some pointers: what arguments should I prepare for ?
> Currently you had "--install" and "--uninstall". So how to specify version/
> revision or "latest" (I suppose only on install) ?
> Something like "--latest" and "--version vv.rr"  ? Or separate version and
> revision like "--version vv --revision rr".

I personally like the --version vv.rr approach more, but I don't mind if the other one is chosen. Also --latest would be fine I think.

**Comment by AV00** [comment]
<style type="text/css">
p, li { white-space: pre-wrap; }
</style>
On giovedì 16 febbraio 2017 11:13:44 CET
Sebastian Bauer wrote:


> I personally like the --version vv.rr approach
more, but I don't mind if the
> other one is chosen. Also --latest would be fine I
think.


I'm working on it and, aside from struggling with
getopt and my limited 
knowledge of bash scripts, I'm seeing a problem with
the above approach: what 
to do if neither --version nor --latest are supplied
?


I'd change it to this: use the version if provided, if
not grab the latest. 
Thus latest would become the default. As of now, I
don't have a default 
(unless you want to force it to use the hardcoded
version you used in the 
original script).


What is it going to be ?


-- 
Bye, A.

**Comment by sba1** [comment]
> I'm working on it and, aside from struggling with
> getopt

getopt is IMO not necessary. See e.g. https://stackoverflow.com/questions/192249/how-do-i-parse-command-line-arguments-in-bash.

> I'd change it to this: use the version if provided, if
> not grab the latest.
> Thus latest would become the default. As of now, I
> don't have a default
> (unless you want to force it to use the hardcoded
> version you used in the
> original script).

Sounds okay to me.

**Comment by AV00** [comment]
On sabato 18 febbraio 2017 05:40:26 CET Sebastian Bauer wrote:
> > I'm working on it and, aside from struggling with
> > getopt
> 
> getopt is IMO not necessary. See e.g.
> https://stackoverflow.com/questions/192249/how-do-i-parse-command-line-argu
> ments-in-bash.

Well, I agree with the sentiment and I left getopt there just because I found 
it. In other words, I believed you thought it to be necessary. ;-)

> Sounds okay to me.

All right, so, when you'll need to build using a specific SDK, pass the 
arguments as "--install --version 53.24" or whatever you like.
I'm finishing the script now and upload it later today or tomorrow.



-- 
Bye, A.

File: packaging/deb/adtools-sdk/update-adtools-sdk-nonfree [hunk 1]
```diff
@@ -22,25 +35,42 @@ display_usage() {
...
+
+		id=$(echo "$CODES" | find_id $target_ver $target_rev)
+		FULLURL=$BASEURL$DOWNLOADURL$id
 		wget "$FULLURL" -O ${EXTRACTDIR}/$OUTNAME
 		lha xw=${EXTRACTDIR} ${EXTRACTDIR}/$OUTNAME
 		lha xw=${EXTRACTDIR} ${EXTRACTDIR}/SDK_Install/base.lha
 		lha xw=${EXTRACTDIR} ${EXTRACTDIR}/SDK_Install/clib*.lha
 		lha xw=${EXTRACTDIR} ${EXTRACTDIR}/SDK_Install/newlib*.lha
+		exit;
 		mkdir -p ${CROSS_ROOT}/ppc-amigaos/SDK/include
```
Comment (sba1) [inline packaging/deb/adtools-sdk/update-adtools-sdk-nonfree]: Why this exit here? Did you forget to remove it?

**Comment by AV00** [comment]
On Thu, February 23, 2017 10:19 pm, Sebastian Bauer wrote:

>
Why this exit here? Did you forget to remove it?

Yes, sorry.
I've fixed it now.



> 
> --
> You
are receiving this because you were mentioned.
> Reply to this
email directly or view it on GitHub:
>
https://github.com/sba1/adtools/pull/22#pullrequestreview-23589484

**Comment by sba1** [comment]
I applied your changes with some modifications. See commit ba33d3297e54081c206bafdd0eca7042acf74283. I still pin the SDK to 53.24 but I will change this as soon as possible.

**Comment by sba1** [comment]
Thanks for the contribution!

**Comment by sba1** [comment]
There will be an adtools-sdk-53.30 available that will install the current SDK.

**Pull request closed without merging**